Rearrange F5 'debug' lines for convenient use in normal game play

JRottm opened the pull request:
During normal game play some of the 'debug' output is actually often
indispensable:
- current coordinates (Where are my friends? Where are my bones?)
- current direction
- node pointed at (What's this funny plant? Is that silver, tin or mithril?)

But to see this you have to enable 3 lines of stuff most of which a normal
player doesn't care about (e.g. map seed). Together with a busy chat it can
make quite a mess of text.

Solution: the 3 essential things listed above easily fit in 1 (one!) line,
together with the MT version (we do want to see it if someone sends bug
reports with screenshots).

The compass direction knows about NW, NE, SW, SE now, the angle is rounded
to full integers. The X/Y/Z coordinates, too, unless show_debug is enabled.
The reason for that: if you build a house and make one wall up to X=10.4
and the other up to X=10.6 your walls will have different lenghts, because
you should have rounded 10.6 to 11 in your head. Now MT does this for you.

The rest is mostly performance related (FPS, drawtime, dtime_jitter, RTT)
and fits all in the 2nd line, which is now only enabled if F5 is pressed
one more time. The map seed isn't performance, but I put it in line 2
anyway. To accomodate all I merged range_all into view_range (and dropped
the decimal place). The selected node's param2 stayed with the node type,
but is only shown if line 2 is visible, normal players don't need it.

To address @paramat's concerns, if the screen width appears too small the
MT version, map seed and pointed node's param2 (i.e. the non performance-
related items) get 'wrapped' to an additional middle line, enlarging the
debug info from 2 to 3 lines again on smaller screens. (Of course the chat
moves to follow suit.)

If only the status is enabled (for normal players) then it's still only 1
single line, which is shortened by hiding the MT version string.

The info text ("Furnace out of fuel") is (permanently) moved down so it can
no longer overlap with the bottom-most chat line (as it could before this
patch series).

Fixer-007 commented:
Make sure chat is displayed below this...

JRottm commented:
I'm not sure I understand what your question is, but the chat is of course displayed directly below the 0, 1 or 2 lines of debug output.

Fixer-007 commented:
Last time we rearranged it, there was bug with chat, that is why I remind %)

JRottm commented:
I tested this, works fine. But thanks for the warning. :-)

paramat commented:
Some of this makes sense.

> the 3 essential things listed above easily fit in 1 (one!) line,
together with the MT version (we do want to see it if someone sends bug
reports with screenshots).

I assume you're testing with default screen width of 1024.
Seed is also essential when taking a screenshot of something, so that someone else can recreate that world and find the same environment. So MT version is better displayed with seed.

> The rest is mostly performance related (FPS, drawtime, dtime_jitter, RTT)
and fits all in the 2nd line, which is now only enabled together with the
profiler graph.

The profiler graphs obstruct the screen though, it is useful and necessary to see the compact readouts of FPS, drawtime, dtime_jitter, RTT above an unobstructed screen.

> To accomodate all I merged range_all into view_range.

Good idea, i should have done this recently.

So the current behaviour has reasons and needs to be preserved, so i feel it's better to add another F5 stage, a new first stage that displays co-ords, direction and pointed node only, then keep the later F5 stages unchanged.

JRottm commented:
I considered another F5 level, but didn't do it because with wireframe there are 4 already. But ok, if you think it's better I'll add a 5th level.
But I would really like to keep the MT version in the 1st line, e.g. then you can see what people are using in YouTube videos. Hmmmmm... not sure about that yet.

JRottm commented:
I'll call the new config option "show_status", any better ideas? "show_position" would be misleading as direction and pointed node are also included...

Fixer-007 commented:
I wonder if F5 debug info can be made configurable, for example:
in minetest.conf
```
debug_info_line1 = coordinates, fps, drawtime, pointednode, collisionboxes, blahblah
debug_info_line2 = blabla, other
debug_info_line3 = blabla
```

JRottm commented:
@Fixer-007: of course, if you find someone willing to implement it. :-P  IMHO it's a bit overkill and I'm already occupied with the changes @paramat asked for, so sorry from me. :-)

JRottm commented:
Almost done, will finish tomorrow...

SmallJoker commented:
Current state of development:
![grafik](https://user-images.githubusercontent.com/1497498/29109050-dea928c6-7ce1-11e7-9aba-fcfd16da3eca.png)
Whereas the next hit on F5 would show the profiler, but no new debugging lines on top.

paramat commented:
Thanks. Please can the @ be replaced with 'at'? It's only 1 character longer.
'param2 =' is now unclear as to what it is about, perhaps 'pointed param2 = '?

paramat commented:
How does 'show_status' behave and what is it for?
Default should be no information, then each F5 press cycles through the other 4 displayes.

None -> position and pointed -> 3 lines -> 3 lines plus graphs -> wireframe -> none

paramat commented:
Perhaps 'param2' should stay grouped with nodename? It looks a little lost anywhere else.
`pointed = (default:dirt, param2 = 0)`

JRottm commented:
Actually won't have much time today...

What SmallJoker hasn't shown is that it switches back to 2 lined if screen width >= 1280.
2 lines make me happy, 3 lines make paramat happy ==> everyone happy!!!  :-)  :-) Yay!

> @ be replaced with 'at'? It's only 1 character longer

It's not about length. In fact we use a proportional font, '@' is wider than 'at', look at it. My idea was to help the eye find quickly where the node name starts in this line of text. Yes, I did think this through!

> 'param2 =' is now unclear

I know, I thought so, too. WIP. (Also Fixer-007, yes, chat isn't moved down properly yet.)

> 'show_status' ... what is it for?

Show_status enables the top line: position, direction, node; show_debug enables the debug output, dtime_jitter and whatnot. They're seperate F5 steps now after all. Normal players use show_status, developers enable both.

How would people feel about no 'varname =' at all in the Status line? Maybe something like this:
**Minetest 0.4.17: (43, -30, 143), 178° South [-Z], default:stone_with_coal**
Comments everyone?

paramat commented:
> My idea was to help the eye find quickly where the node name starts in this line of text.

Ok good point and it does help with that, but it also seems like a sloppy internet-slang type abbreviation, but lets see what others think. I think 'Pointed = ' is clear enough already.

So 'show status = true' makes the 1st line displayed by default? I think many or maybe even most players don't want this enabled by default, instead they can show it with F5, then we won't need another setting. After all this basic information is currently in F5 and players mostly play without F5 enabled.

At the least 'show status' should be default false, like 'show debug' is. But i don't think another setting is justified, the 'show debug' setting is fairly useless, i didn't even know it existed, we can just press F5 instead, and the setting doesn't cycle through the F5 steps either.

paramat commented on src/game.cpp:
```diff
@@ -995,14 +995,16 @@ static void updateChat(Client &client, f32 dtime, bool show_debug,
+	if (flags.show_debug) {
+		chat_y += 2 * line_height;
+	} else if (flags.show_status) {
+		chat_y += 1 * line_height;
```
Nice, i agree with having a 'show status' flag to determine how far chat is moved down, but i think it should just be an internal variable not settable with a setting.

paramat commented on src/game.cpp:
```diff
@@ -2870,40 +2873,39 @@ void Game::toggleFog()
+	// Initial / 5x toggle: Status text with chat
+	// 1x toggle: Status and debug text with chat
+	// 2x toggle: Status and debug text with chat and profiler graph
+	// 3x toggle: Status and debug text with chat and wireframe (skipped if no debug privilege)
```
Please let's keep behaviour consistent with current, default being chat only.
```
+	// Initial / 5x toggle: Chat only
+	// 1x toggle: Status text with chat
+	// 2x toggle: Status and debug text with chat
+	// 3x toggle: Status and debug text with chat and profiler graph
+	// 4x toggle: Status and debug text with chat and wireframe (skipped if no debug privilege)
```

JRottm commented:
**Whew** that took a while, had to do 2 refactoring pre-patches before I could implement *cleanly* what I had in mind. See the respective commit messages for what they do.

And the main patch has also changed a bit: the direction knows about NW, SE, ... now, the position is rounded to full nodes unless show_debug is on, etc. I didn't remove the "varname =", though.

I'll **edit** the initial PR description to match the current patch series. And i'll try to make screenshots, if I can manage.

JRottm commented:
![screenshot](https://user-images.githubusercontent.com/30634967/29168555-ab2c33c8-7dcf-11e7-9fc6-7ef8f774401c.png)
This is how it would look with 1024x... and 1280x... resolution with show_status enabled alone (normal player) and with show_debug (developers).

JRottm commented:
@paramat 
> 'show status' should be default false, like 'show debug' is

In fact show_debug defaults to true #ifndef NDEBUG. IMHO this makes sense, some developers don't like having to press F5 every time.

And I as a player wouldn't like to have to press F5 every time to see my coordinates; I, personally, had the coords enable **all the time** when playing, even though all this debugging mess around it bugged me a lot. And now that looks much cleaner and is only 1 line, I think a majority of players would find it convenient to have the status enabled.

Please, comments, what do other players think?

paramat commented:
> In fact show_debug defaults to true #ifndef NDEBUG.

Yes i know, but that's for debug build, a special case.

> And I as a player wouldn't like to have to press F5 every time to see my coordinates; I, personally, had the coords enabled all the time when playing

Ok, but there will be a setting (assuming we keep that) to configure behaviour to what you like. Minetest has always had position information off by default and i can't remember anyone complaining about that.
I believe most players want the clearest most text-free screen as the default and not be forced to see information they may not want, this is how it has been and it makes intuitive sense. Co-ordinates, direction and pointed node are not needed at all during most playing time.

Position:
A precise position is useful and i feel should be displayed precisely whenever position is seen, the decimal place makes the displayed position match the fundamental Irrlicht co-ordinate unit. I disagree with assuming that because a player is on the first F5 stage they want a rounded position, there's no reason at all to round it and it's just a little more processing.

Direction:
The added compass directions may not be that useful, as Minetest worlds are cube-based and based on the 4 primary directions. I often use the direction readout as a clear way to tell me if i am pointing 'more north than east' or 'more west than south'. Many aspects of Minetest depend on which of the 4 directions you are facing, not which of the 8 directions you are facing, so this change is a little inconvenient. The current behaviour clearly shows you when you are passing 45 or 135 or 225 or 315 degrees.

I still feel the '@' symbol looks like sloppy internet slang, although i understand why you did it. Using the word 'at' will not be significantly less clear. Whenever i see '@' used as a replacement for 'at' it does not give a good impression, it's lazy internet abbreviation like SMS language.

So for multiple reasons i am :-1:

paramat commented:
To clarify, i feel exposing 'show status' as a setting is not much needed but i'm ok with that if other devs want it.
What i am requesting is:
F5 information is disabled by default.
Position is not rounded.
Change the '@' to 'pointed = default:dirt', the '=' catches the eye in the way you intended with the '@', but it is consistent with all other data, which use a '='.

paramat commented:
> This is how it would look with 1024x... and 1280x... resolutions with show_status enabled alone (normal player) and with show_debug (developers). You'll notice I tried to spread the info evenly across the visible lines to make them as short as possible as a whole.

I disapprove of this game-window-dependent feature too, for 3 reasons:
1. With a larger game window, the information lines take up less space relatively already, so reducing 3 lines to 2 is not necessary.
2. Now the information in each line depends on game window size, and moves around depending on that. With F5 information, players get used to where each value is, and need to do this to quickly see the information they want to see, having values move around when moving to fullscreen and back, or when altering your game window, will be disorientating. Minetest version actually moves to a different F5 stage depending on game window size, this is bad and it needs to stay on the same stage as seed for the reason i explained above.
3. This feature makes this area of code unnecessarily complex, the core philosophy of Mietest is simplicity, and with our severe lack of core devs we need to avoid unnecessary complexity.

So, my request is: please don't make any changes to behaviour with a large game window size.

JRottm commented:
I disagree - obviously, otherwise I wouldn't have done things the way I have. You rejection isn't about code quality, mostly just personal taste.

> F5 information is disabled by default as currently.

Formerly this involved a mess of debug output, the situation has changed.

> Co-ordinates [...] not needed at all during most playing time

I had to have it enabled all the time, for TPs, to find my base or bones, despite the other debug clutter being very distracting. That's why I did this PR in the first place.

Unfortunately noone else has shared any opinion, so I comply. I disabled it the way you like it.

> '@' used as a replacement for 'at' it does not give a good impression

I do like the @, because, as I explained, it helps the eye catch quickly where the node name starts. But I changed it back to 'at', as you prefer.

I removed the '=' after pos and yaw, too, this should address your consistency concern, and I find it looks friendlier, less technical. The status line is meant for use by normal players, whereas the rest of the debug output is for developers, the audience is different.

> A precise position is useful and i feel should be displayed precisely whenever position is seen, the decimal place makes the displayed position match the fundamental Irrlicht co-ordinate unit.

It's a float, all decimal places change with every movement. Do "std::setprecision(8)" and you'll see.
If you want precision, shall I change it to 8 digits for debugging?

> [...] round it and it's just a little more processing

Both setprecision(0) or (1) are rounded.

> I disagree with assuming that because a player is on the first F5 stage they want a rounded position [...] I feel players want the same form of position whenever it is seen.

I **am** a player and I feel rounding is a big improvement. Decimal places tell me if I'm in the middle of the node. I don't need that! I can see that just by looking, and, if it matters, I sneak. What I do need is to know where I am. And the decimal places are very distracting when I need to know exactly which node I'm standing on. Quote from my PR description: if you build a house and make one wall up to X=10.4 and the other up to X=10.6 your walls will have different lenghts, because you should have rounded 10.6 to 11 in your head. I'm not **assuming**, I test-played my change, and the GUI rounding for me is a relief. I thoroughly consider every change I make.

> I disapprove of this game-window-dependent feature too [...]
With a larger game window, the information lines take up less space relatively already, so reducing 3 lines to 2 is not necessary.

I see this from the opposite direction: what made the current behavior so terribly distracting was this mess of 3 lines debug output + 6 lines chat, it was so hard, at a glance, to see where the chat began. For this reason I wanted status/debug to use as few line as possible. ==> The 2 lines were my primary goal. The 3 lines are only an unavoidable necessity, because you correctly warned that some people might have very small screens.

If 2 lines is **primarily** what I want, but I can't avoid having to support shorter lines, too, then some items need to be wrapped.

> having values move around when moving to fullscreen and back, or when altering your game window, will be disorientating.

Admittedly, a little, but I think people can handle. Most people won't change their window size all the time. I tried to keep things logical and in the same order.

> This feature makes this area of code unnecessarily complex,

Good point. I did try to write clear and clean code. You will notice that I even split off two(!) preparatory refactoring patches in my patch series to clean the area before the actual changes. Each of the 4 patches in my series can be compiled and run to allow bisecting. My commit messages are much more detailed than most I can see in "git log". And to counter your worries I have today added many further comments to this area. I think the code is clearer than it was before my PR.

> the core philosophy of Mietest is simplicity,

This reasoning could be applied to every feature ever. Not counting all the comments I added, the new game.cpp is just 9 lines longer than the old one.

> and with our severe lack of core devs [...]

I like MT, and I want to contribute my little part to make it better, and yet I have spent **significantly** more time on defending the motivation behind my features than on actually coding them.

sfan5 commented on src/game.cpp:
```diff
@@ -1429,16 +1419,16 @@ class Game {
-	gui::IGUIStaticText *guitext_status;
-	gui::IGUIStaticText *guitext_chat;	   // Chat text
+	gui::IGUIStaticText *guitext_info;     // Object info (e.g. "Furnace out of fuel")
+	gui::IGUIStaticText *guitext_notice;   // GUI messages (e.g. "Minimap hidden")
```
while it might be a clearer, it adds too many pointless changes to this commit/PR IMO

JRottm replied:
The old comment was downright misleading, guitext_info isn't the one "at the middle of the screen", it's at the left. It did take me a while to understand which text does what, so I wanted to make it easier for the next guy...

BTW, I split my changes in 4 separate patches, they're probably easier to read separately.

rubenwardy replied:
this is fine imo

paramat replied:
Fine for me apart from the 'status'-> 'notice' change.

sfan5 commented:
I generally like this :+1:
Just two things:
* Is it new that the `Minetest <version>` string is no longer shown by default?
* On small screens why not move all info relating to the pointed node to 2nd line and keep mt + ver on the first?

rubenwardy commented:
I'm not sure about adding yet another stage to get to the graphs.

JRottm commented:
@rubenwardy: I hadn't done this in the initial version of this PR, because I didn't like it either, but @paramat explicitly asked me to.

JRottm commented:
@sfan5:

> Is it new that the Minetest <version> string is no longer shown by default?

Before this PR:
MT version is part of debug info i.e. only shown if show_debug=true. Default of show_debug is false unless #ifndef NDEBUG.

After this PR:
MT version is part of basic status i.e. first F5 stage, shown if show_status=true (at least on larger screen, on small screen sacrificed, because player pos, direction and pointed node are more important).

I wanted show_status=true per default, but paramat insisted it to be default off.

(Oh, #ifndef NDEBUG show_status should be likewise true, I'll change that! New push coming up!)

> On small screens why not move all info relating to the pointed node to 2nd line and keep mt + ver on the first?

You mean in the status+debug both shown case? Or always?
For me player pos, direction and node name are the most interesting during normal play. (See PR description: "What's this funny plant? Is that silver, tin or mithril?")

sfan5 commented:
> You mean in the status+debug both shown case? Or always?

In the status+debug case which is the only case (on small screens) where the Minetest string would show up anyway.

JRottm commented:
**Pushed update!** Now show_status defaults to !NDEBUG.

@sfan5: So on small screens, when you press F5, MT version would appear at the start of line 1 whereas pointed at would get pushed to line 2.

Hmm, not too fond of that for 2 reasons:

1) Currently when pressing F5 1st line remains the same, only 2 more lines get added. Paramat already criticised me for moving items when the window size changes and I tried to defend myself that people don't change their window size all the time. But even I would prefer items not moving around all the time whenever F5 is pressed, because that will happen more often.

2) Grouping: pos, direction and pointed node belong together, they're my "basic status". Whereas MT if no room in line 1 fits well together with seed, because MT ver + seed are the 2 things needed for exactly reproducing a world. (In large screen case they're also together -  horizontally.)

paramat commented:
> You rejection isn't about code quality, mostly just personal taste.

No, it's for practical reasons that affect all players, as explained.

> I removed the '=' after pos and yaw, too

The '=' helps with readability much like your intention with the '@', otherwise it becomes solid text.

> It's a float, all decimal places change with every movement. Do "std::setprecision(8)" and you'll see.
If you want precision, shall I change it to 8 digits for debugging?

I meant, please leave it as it is, 1 decimal place wherever position is shown.

> I am a player and I feel rounding is a big improvement. Decimal places tell me if I'm in the middle of the node. I don't need that!
And the decimal places are very distracting when I need to know exactly which node I'm standing on

Normal players find the decimal place useful for many reasons, and are used to it, and have not complained. Maybe you're new to MT and haven't got used to a few things yet.

> I wanted status/debug to use as few line as possible. ==> The 2 lines were my primary goal. The 3 lines are only an unavoidable necessity, because you correctly warned that some people might have very small screens.
If 2 lines is primarily what I want, but I can't avoid having to support shorter lines, too, then some items need to be wrapped.

Good intention, but a game window < 1280 is not 'very small' and a rare situation, for MT players that is the majority usage.
I have a fairly powerful desktop and play with a 1200 width screen.

> After this PR:
MT version is part of basic status i.e. first F5 stage, shown if show_status=true (at least on larger screen, on small screen sacrificed, because player pos, direction and pointed node are more important).

As above, < 1280 is the vast majority of MT players, you should consider that the standard usage.

sfan5

> On small screens why not move all info relating to the pointed node to 2nd line and keep mt + ver on the first?

Because it's bad to move things around depending on screen size, it's unnecessary code complexity that results in confusing inconsistent behaviour.

paramat commented:
Thanks for the changes.

My current requests are:
Position remains rounded to 1 decimal place in all stages.
Change the 'pointing at' to 'pointed = ', the '=' catches the eye in the way you intended with the '@'.
Re-add '=' to all fields to show where values begin and to be more readable.

I'm not keen on changing number of lines and moving values around due to screensize, but if other devs like this then i'll have to accept that.

JRottm commented:
> My current requests are: Change the 'pointing at' to 'pointed = ' [...] Re-add '=' to all fields

I'm sorry, but no. I think it looks much friendlier this way, like a polished game. '=' looks nerdy. Look at the text "Viewing range changed to 120" for example. It doesn't say "New view_range = 120". IMHO the status output is part of (non-technical) players' user interface. The remaining debug output is for developers, and it *does* in fact say "view_range = 120" there.

> 1 decimal place wherever position is shown. Normal players find the decimal place useful for many reasons, and are used to it, and have not complained.

*Which* "many reasons"? You argued for 1 decimal place to "match the fundamental Irrlicht co-ordinate unit", but normal players have no clue what that even means. I test-played my change and find it's a big improvement, and I've explained the reason in my PR description. "not complained" - well, I *did* make this PR changing it, and I'm typing these lines defending my changes... So far your reasoning doesn't convince me at all, so I'm gonne stick to my change.

> Maybe you're new to MT and haven't got used to a few things yet.

??

> a game window < 1280 is not 'very small' and a rare situation

That's ok, my layout works fine without even overlapping with the minimap on 800x600, whereas 0.4.16's text gets cut off hard already at 1024x768.

> < 1280 is the vast majority of MT players, you should consider that the standard usage.

In fact my 2-line layout would work from about 1100 pixels upwards - with default font settings. I didn't know if some people use larger fonts, so I added a large safety margin. (Dynamic detection would've been too complex.)

Do you think it would be safe to switch to 2 lines earlier?

JRottm commented:
**Pushed update!** Forget my last question, I found a way to switch depending on font size without adding complexity or CPU. It's not perfect, but it's much better than hardcoding it.

paramat commented:
I'm now ok with rounding to integers in the first F5 stage, those who want the decimal place can use the 2nd stage.
The '=' is not a big issue, so removing my -1.

JRottm commented:
**Rebased** to current master tip to resolve conflicts with @nerzhul's code modernization.

Please would someone move this PR forward... please...??

nerzhul commented:
@JRottm sorry i'm poluting with modernization but it's nearly finished :p

nerzhul commented on src/game.cpp:
```diff
@@ -4297,133 +4281,188 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
+		if (flags.show_debug) {
+			static float drawtime_avg = 0;
+			drawtime_avg = drawtime_avg * 0.95 + stats.drawtime * 0.05;
+			unsigned int fps = myround(1.0 / stats.dtime_jitter.avg);
```
u32

nerzhul commented on src/game.cpp:
```diff
@@ -4297,133 +4281,188 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
+	/*
+		Update GUI notice
+	*/
+	float noticetext_time_max = 1.5;
```
1.5f

JRottm commented:
**Pushed update** fixing the data types @nerzhul pointed out and other similar cases in the code my patch series touched.

Wuzzy2 commented:
While I'm OK with reordering stuff, I think the general promise of this issue is misguided.

It is a conceptual mistake to conflate **debug information**, which is somethig for developers with **player information**, which is something for players.
I think the debug screen originally was never meant to be used/abused by players in the first place. Therefore, trying to make it more *player*-friendly is basically solving the wrong problem.

I have argued before that player information and debug stuff should be kept strictly separate so the player is not forced or temped to abuse the debug screen. See here: https://github.com/minetest/minetest/issues/4556

Just a quick recap why it is a bad idea to give out these information **for free**:

> current coordinates (Where are my friends? Where are my bones?)
> current direction

This destroys a very interesting gameplay mechanic in which you have to first *earn* these information by e.g. crafting tools. Like in the orientieering mod (craft tools to see coordinates, yaw, etc.).

> node pointed at (What's this funny plant? Is that silver, tin or mithril?)

This destroys invisible trap mods.

Oh, and by the way: Please add `param1` back in to the debug screen. Why has this been removed? It didn't hurt anyone. We have `param2` already in, so not having `param1` seems strangely inconsistent.

Wuzzy2 commented:
Oh, if you already work on debug screen, please take this issue into account, too:
https://github.com/minetest/minetest/issues/5920
(it has been closed too early)

JRottm commented:
@Wuzzy2: I just regroup/format the existing info to make the screen less messy. I don't add or remove **any** F5 items, that's beyond the intended scope of this patch series.

> mistake to conflate debug [...] with player information

That's why I did separate them: show_status (line 1, 1x F5) vs. show_debug (lines 2+3, 2x F5).

> should be kept strictly separate so the player is not forced or temped to abuse the debug screen [...] #4556

I decidedly oppose that.

1) I hate it when SW patronizes me, telling me how I'm *supposed* to enjoy my game. If I prefer minimap and position shown all the time then it's not the game's business telling me that's not how to do it. MT is open-source open-world and IMHO should be the exact opposite of Apple's and MS's treat-users-like-children mentality. Even if it were some labyrinth map, that's **my** decision. Detective novel books don't prevent you skipping to the last chapter, either. The only exception is to guarantee fairness when *competing against* other players (as opposed to merely playing *with* others, sharing a world).

2) It hinders debugging. Bugs don't appear when you expect them, if a problem occurs during normal gameplay it's no use telling the player, please quit, then change a config setting, then reenter the game and then press P for a trace.

That being said, my patches actually aid you in your goal, because they separate 'spoiler' info (position, ...) from debug (FPS, ...). A further PR could then be made to block show_status if the game forbids it (like it can already be done with the minimap) whereas show_debug would stay accessible. My display code in updateGui() already does work if show_status=false, show_debug=true, even though this flag combination cannot occur 'naturally' by pressing F5.

octacian commented:
Personally, I'd like to see the f5 debug split into a first mode, which displays only FPS, coordinates (configurable), and yaw (configurable); a second mode, which displays the above along with general client-side performance related information (e.g. `drawtime`, viewing range, seed, etc..); and a final mode that displays all the other more sensitive information (like the pointed node) and requires that the player has the `debug` privilege (as does the wireframe mode).

Items I suggested to be configurable should be done via player HUD attributes (e.g. `debug_coordinates`, `debug_yaw`).

paramat commented:
```
 Refactor "*status*" to "*notice*" in game.cpp preparing for F5 change

In the upcoming commit we will introduce a show_status flag displaying the
current position, direction and node pointed at. I called it "status",
because it permanently displays essential information, like a "Status bar"
in many other programs.

However game.cpp already has a text named "status", it shows messages like
"HUD hidden". But these aren't permanent 'status', they notice the user of
GUI changes. Thus, to avoid confusion, rename all related variables:
```
I'm fine with the actual F5 changes, but disapprove of this refactor commit.

You have called the 1st line of F5 information 'status', which is not a particularly good word for it, it's vague, more descriptive would be 'position', 'location', or better still 'orientation' which implies position, direction and the player's relation to the surroundings (node pointed at).

Because you have chosen 'status' you have forced yourself to now change a lot of code to change 'status text' to 'notice text', although 'notice' is slightly more accurate than 'status' this is best left alone unless there is a very good reason for a change, which there isn't.

I think it's better to split off the refactor into a later PR so we don't feel pressured to merge it with these useful F5 changes. Then also use a different word for the 1st line of F5 information.

So i'm -1 for the refactor and using 'status' for the 1st line.

paramat commented:
Maybe 'location' is best, since that implies position but also what your surroundings are (pointing at).

Wuzzy2 commented:
> That's why I did separate them: show_status (line 1, 1x F5) vs. show_debug (lines 2+3, 2x F5).

That's not a real separation, it's still the same debug mode key and the critical information is just 1 keypress away, so it's easy to get “spoiled” by accident.

> I hate it when SW patronizes me (…)

> If I prefer minimap and position shown all the time (…)

This all breaks down to 2 types of gameplay: Survival vs Creative Mode. Both are legitimate.
In Creative-style gameplay, you are correct in all aspects. Here it's really “anything goes” and any artificial restriction doesn't make sense.
But there's also “Survival”-style gameplay, when Creative Mode is off. The point of this is the challenge. Players need and WANT to be restricted here, otherwise there's no challenge at all, without restrictions it's just another Creative Mode. That's why we don't enable fly mode, fast mode, etc. by default. Plus, if you use the orientieering mod, it takes restrain to not just press F5 and you might still hit it by accident and get “spoiled”.

It's not about forcing anyone to do anything at all. My goal is more to make Survival-like gameplay *possible*.

> A further PR could then be made to block show_status if the game forbids it (like it can already be done with the minimap) whereas show_debug would stay accessible.

Yes, this was my point all along. Although it seems strange that they share the same key.

> It hinders debugging. Bugs don't appear when you expect them, if a problem occurs during normal gameplay it's no use telling the player, please quit, then change a config setting, then reenter the game and then press P for a trace.

I am not really opposed to the “Print Stacks” key.

JRottm commented:
@paramat: 12(!) days ago I proposed "show_status" and explicitly **asked** for better suggestions. Nobody spoke up.

For me show_orientation implies yaw, I'd never expect it to display the "pointed at" node. show_position and _location imply, well, position i.e. are just as misleading. "show_status" is certainly not perfect, but I still find it much clearer.

> although 'notice' is slightly more accurate [...] i'm -1 for the refactor

nerzhul just overhauled the whole tree to make it cleaner. That's a good thing, everybody was fine with that. I just want to rename a few variables local(!) to 1 file, what's the big deal? 8 days ago you criticised me that by adding features I'd add complexity, yet today you -1 me for making the code more readable?? I even split off the renames into a separate patch to ease review or possibly bisecting and my commit message lists every variable's old and new names. What else could anyone want?

JRottm commented:
@Wuzzy2: I thought about adding a dedicated key for the status, but as you noticed yourself (#6170) we have so many keys already you can't even configure them all on 1 screen. :-)

> when Creative Mode is off [...] Players need and WANT to be restricted

I agree, but I alone want to decide what these restrictions are. If I want to survive against mobs/starvation, but still find my house with F5, that's mine to decide.

Exception, as I said, when competing against others. But in this case people could patch their clients to still show the position. And if you can't prevent cheaters from doing it, perfect fairness can only be achieved by allowing it for honest players, too.

> it takes restrain to not just press F5

No problem :-)

JRottm commented:
@octacian: I did consider including FPS in the status line, but in the end I decided it would fit better with the other performance-related stuff, like dtime and RTT.

Ultimately, maybe @Fixer-007's suggestion (12 days ago) of a fully customizable status line(*) may be the only way to give everyone exactly what they wish, but that's well beyond what my PR was intended to achieve.

*) idea: printf-like format strings:
`pos (%.0c), yaw %.0y° %Y [%X]%(, pointing at %p%)`
%c would be coordinates (or "unknown" if forbidden by HUD flag), %y = yaw as number (or "unknown"), %Y = yaw as string ("Northwest"), %X = coordinate change as string ("+X-Z"), %p = pointed-at node name, %( and %) would mark substring to skip if none of the expressions inside (%p in this case) was valid, %% would be literal % as usual, you could easily have 52 different more or less useful variables at your disposal

People could have as many decimal places, "=" or "@" signs as they liked. But I wouldn't have to implement this -- *correction* -- wouldn't have to get this through review. Implementing would be 3 days, but my simple PR is stuck in review for 2 weeks already, without a single approval yet, this printf thing would take 5 months of bikeshedding to get merged...

octacian commented:
> Implementing would be 3 days, but my simple PR is stuck in review for 2 weeks already, without a single approval yet, this printf thing would take 5 

You know, if I were you, I'd go for it, although I'd suggest you do it in a separater PR. The PR being stuck in review isn't necessarily *bad*, although it's not *good* either. The devs can either be very busy or find the feature to be something that needs lots of discussion.

paramat commented:
> Oh, and by the way: Please add param1 back in to the debug screen. Why has this been removed? It didn't hurt anyone. We have param2 already in, so not having param1 seems strangely inconsistent.

Wuzzy and i have discussed this now and i have explained why it was removed.Pparam1 is almost completely useless as it is 0 for all solid nodes and unreadable for all light sources due to lightbank encoding. Air nodes cannot be pointed at so light value for air nodes cannot be seen. I believe Wuzzy agrees now it is not justified.

> 12(!) days ago I proposed "show_status" and explicitly asked for better suggestions. Nobody spoke up.

Yes, that's MT dev for you, things will be ignored to then later people speak up.
'Orientation' or 'location' are much more suitable then 'status' which could mean anyhting.

> I just want to rename a few variables local(!) to 1 file, what's the big deal?

I have explained, i just have an opinion that unnecessary refactoring caused by choosing a not very good name 'status' is possibly not a good idea. 'Just wanting to rename a few variables' is you trying to make it sound trivial but this is code, every small change will be reviewed and considered and suggestions will be made, there's no point getting irritated by this or arguing it is unreasonable reviewer behaviour..

> yet today you -1 me for making the code more readable?

I'm not disapproving for readability reasons, your changes do not improve readability.

> What else could anyone want?

Well, i have explained what i want. 'What else could anyone want?' implies you thnk your PR is perfect but reviewers will have opinons about it and will request changes, that's just how dev works.

You seem to be getting overly dramatic and irritated by a standard dev process and reviewers requesting changes. This is what MT dev work is like, it's hard work and frustrating.

paramat commented:
> I agree, but I alone want to decide what these restrictions are. If I want to survive against mobs/starvation, but still find my house with F5, that's mine to decide.

No that's up to the server to decide. MT design is that the server has control over anything that could give a player an advantage or anything that affects the server or server gameplay (See CSM discussions).

> Exception, as I said, when competing against others. But in this case people could patch their clients to still show the position. And if you can't prevent cheaters from doing it, perfect fairness can only be achieved by allowing it for honest players, too.

This is the invalid argument that because hackers can cheat, we shouldn't attempt to stop cheating in any way, the argument has been used in the discussions about restrictions on CSM (see #5915 ) by those who are for some reason resisting restrictions (these people are often hackers themselves or believe they have the right to disrupt servers with their cheating).

paramat commented:
> a fully customizable status line(*) may be the only way to give everyone exactly what they wish,

That would not be justified, it would be overkill.
We could possibly add HUD flags to control what F5 stages a player can use.

paramat commented:
'show status' is not a good choice and not descriptive, 'status' can mean almost anything. Something like 'position', 'location', 'orientation' or similar is descriptive of the first stage of F5.

'To position yourself' can certainly mean direction too.
What i was thinking about 'location' is that the word also means 'the surroundings', which therefore includes 'pointing at'.
My thinking with 'orientation' is that is suggests the sport of 'orienteering' which is very much about position, direction and the surroundings, so also includes 'pointing at'.

Refactors need a good reason to be done, and here this is mostly being done due to the bad choice of 'show status'.
'statustext' is fine and 'noticetext' is no better, so there's no good reason for the refactor.

I can review and possibly +1 this if you remove the refactor and change 'show status' to something more descriptive.
Curently i am :-1: for the whole PR for the reasons above.

paramat commented:
>  If I prefer minimap and position shown all the time then it's not the game's business telling me that's not how to do it

In singleplayer you will have all privs so you can see whatever information you want, but when on a server it is the server that decides what information you can see, if that information affects server gameplay.

> Even if it were some labyrinth map, that's my decision. 

Not on a server, a server would obviously need to disable players seeing co-ordinates or direction or pointed nodes in some situations, and i support adding controls that can control what information can be seen.

> The only exception is to guarantee fairness when competing against other players (as opposed to merely playing with others, sharing a world).

Yes, the server decides in which worlds information needs to be controlled. In many worlds, such as creative worlds, the F5 information is no issue.

paramat commented:
Wuzzy2

> It is a conceptual mistake to conflate debug information, which is somethig for developers with player information, which is something for players.

Separating these 2 is the mistake, players find 'debug information' very useful, and developers use 'player information' all the time. Many MT players are modders or contributors so are developers, and developers play the game while developing (you can only develop a game by playing it), there is no clear division.
If we separate information into different F5 stages, it is best separated for other reasons

paramat commented:
octacian i do agree with HUD flags to control what information can be seen, as long as it is as simple as possible, what i considered overkill was control of every element as described here https://github.com/minetest/minetest/pull/6225#issuecomment-323614489

Wuzzy2 commented:
Paramat: Some players are also developers and some developers are also players. BUT not all players are developers! So your argument against separation is strange.

You seem to fail to understand my bigger point here. I am NOT suggesting to take anything away from the debug screen. In fact, we could expose information in a “player and gameplay-friendly manner” while not touching the debug screen at all. This might seem redundant but this kind of separation is *crucial* to enable certain types of gameplay.

Because your argument does not address at all the point in that current debug screen destroys the purpose of `orienteering` and trap nodes.

octacian commented:
> what i considered overkill was control of every element

It's these "overkill" things that seem somewhat unjustified that often give a program more value from a user's perspective. And, please don't respond about how developers don't have time to deal with these things. That's why we have contributors.

octacian commented:
In regard to paramat's argument against separation: I think that we *should* separate the *player* information from *developer* information, but not in a way that developer F5 stages lose player information. Instead, I think we should simply have one stage that includes player information but *not* developer information, and another stage that contains player information *and* developer information.

paramat commented:
> BUT not all players are developers

I know, i wrote 'players find 'debug information' very useful'.

> I am NOT suggesting to take anything away from the debug screen

I know :]

> Because your argument does not address at all the point in that current debug screen destroys the purpose of orienteering and trap nodes.

I actually support changes to have HUD flags to hide certain information.

> It's these "overkill" things that seem somewhat unjustified that often give a program more value from a user's perspective.

When we judge if something is justified, we consider the player too. The example given was per-element control, which is simply unnecessary and has no benefit for players. We can divide the information into fewer parts.
If a contributor is able to code an overkill solution is irrelevant, we don't want overcomplex code because that is unnecessary extra maintenance in the future.
There is no way to justify unnecessarily overcomplex code, MT would be a nightmare if we weren't careful about this, and simplicity is a core MT philosophy.

>  I think we should simply have one stage that includes player information but not developer information, and another stage that contains player information and developer information.

Yeah good idea.

JRottm commented:
Sorry, RL priorities. Quickly **rebased** to current master, but as I now scroll down past yet another barrage of huge posts from you with more thumbs down I think this PR is going to die. :-(  :-(

Sorry, @paramat, I'm really in a hurry, so just browsing quickly what you've written.

Yes, show_status is kinda generic, that's the point, because the status includes 3 quite different elements for which a *specific* name wouldn't fit (and others might be added, some people asked for FPS). Yes, show_position and show_orientation are more specific - for position and yaw respectively, that's why they're actually very misleading when applied to a status line containing 2 other elements, too.

In short replacing "status" with "orientation" or "position" would make the code worse IMHO, not clearer, that's why I don't change it.

I'm ok with a HUD flag blocking show_status if game authors want it, and my code is prepared for it, even though I'm really not a fan, as I said, and would definitely want to override it in singleplayer. But in any case, it's beyond the scope of this PR. This PR meant to only **rearrange** the already existing info, nothing else.

> This is the invalid argument that because hackers can cheat, we shouldn't attempt to stop cheating in any way

This is absolutely **not** what I said!  MT should absolutely stop cheaters, e.g. bendeutsch's Server side movement work is a necessary thing. But I don't see how one could possibly keep the player position a secret from clients. And as long as there is no other way, you can take away cheaters' unfair advantage by showing the pos to everyone.

> those who are for some reason resisting restrictions (these people are often hackers themselves or believe they have the right to disrupt servers with their cheating)

??? ... this is the 2nd time that I'm not sure which tone to read into your posts.

In any case, RL, won't have any spare time for a while...

Wuzzy2 commented:
Sorry. Yeah, this discussion probably went way too far again. Especially far too off-topic. :-(
We (myself including) must really learn to focus, these endless discussions really drag down productivity. I can totally understand why you feel stressed.

OK, since this is just about the reordering think I just leave it as-is.

LazyJ commented:
@JRottm, as paramat mentioned earlier, sometimes it takes awhile for folks to chime in on a PR. For me it started with a conversation in #minetest-dev, 2017-08-11, starting around [20:04](http://irc.minetest.net/minetest-dev/2017-08-11#i_5040751) about whether or not the debug info should be displayed. I made an inquiry and was directed to this PR which lead to another related PR, #4556. Throw in real life demands and it took two weeks to find my way here. A lot of running around for, basically, one topic. It's frustrating to have to wait so long for responses but unfortunately that's the way these things go, at times.

Ok, so my two-cents _(or more)_ worth:

I too keep my debug info on at all times _(one stage before the graphs are shown)_ except when taking screenshots where I don't want any HUD, chat, or hotbar showing.

I like Fixer-007's idea of having a configurable stage in the F5 rotation. Not every body uses the same info that is already shown in the debug. Such a stage would allow us normies to have just what we use and not clutter up the screen with info we don't use or are unable to act upon.

Adding "NW", "NE", "SW", and "SE" to the cardinal direction makes sense. Minetest is geeky so anywhere geek-speek can be translated into plain-talk is an improvement for players who aren't geeks. In real life, for normies at least, most conversations involving directions don't use yaw degrees. Normal life conversations are more like this: _"Head on north, straight through town til ya git to the grocery store. Hang a left at the light and keep going west til ye run into the ferry."_ Even the weather forecasters on TV and radio use "northwest", "easterly", "south by southwest", etc.

For what debug info I regularly use in Minetest, one line is more than enough. Coords, cardinal directions, and modname:nodename are the most used. FPS and view_range are lower priorities. The rest of the displayed debug info I've just learned to ignore.

**Decimal Points in the Coords**
Decimal points aren't needed or used in normal play or admining on servers I've toured over the years. I like the rounded up/down idea. MT's default, chat based teleport command doesn't use decimals. Neither does ShadowNinja's Areas protection mod nor the WorldEdit mod in their chat command use of coords.

**Map Seed**
Investigating terrain based on a map seed only works when the MT engines' match exactly in their interpretation of the map seed. The older the world is, the less likely the map seed will result in the same terrain; not even the same biome if the world is old enough. My server, LinuxGaming2, was started in April 2013, using MT 0.4.6, mgv6, map seed 7029869889726536560, and the static spawn point is set to 444,42,747 _(note no decimals used for this minetest.conf setting)_. The world has never been wiped. The static spawn point has never moved from the **top of the desert mesa** it was created on. I plugged the map seed into a recently compiled MT 0.4.16-dev, set the mapgen to v6, and teleported _(another place were decimals are not used)_ to the spawn point coords. At those coords I was floating in mid-air above low, grassy hills, some small ponds, and a large snow biome to the east. Definitely not the same biomes and no where close to matching the same terrain as it was generated in LinuxGaming more than four years ago.

The next part is repeated on PR #4556 but is relevant to this PR's discussion too.

A couple years ago I enabled the display of debug info on my server because it is helpful to players _(but not so helpful that it is cheating)_ for the way the game play is done on my server. As I mentioned earlier, coords, cardinal direction, modname:nodename, are the most useful. FPS and view_range are secondary. The rest of the debug info is ignored _(though I would like to be able to comprehend what it is telling me well enough to be able to use it to tweak settings or diagnose problems)_. Being able to toggle through various stages of debug detail levels is ideal for a quick glance at the graphs and metrics and then back to the normal, less detailed info.

**FPS**
If the lag seems to have built up after teleporting to a few places or after gathering and mining for a while, a quick glance at the FPS will provide a clue as to whether the lag is real or because the "R" key was unknowingly pressed, toggling on the full-view range causing the FPS to drop like a stone and motion to be very choppy.

**View_Range**
When flying or setting up for scenery screenshots, increasing the view_range on the fly is done. When walking on the ground or mining, decreasing the view range improves performance for those who have weaker systems. For planning builds and landscaping - create a look-out tower, increase the view_range and work on your plan _(and take screenshots for reference)_.

**Modname:nodename**
LinuxGaming2 uses several custom nodes. For newcomers _(new to the server but familiar with MT)_ it is handy to target a custom node they find while exploring to learn what it is. Then they can run a search for it in the craft guide to learn how to craft it.

I use the modname:nodename on a regular basis in conjunction with WorldEdit when cleaning up messes on the server; there are too many nodes to remember to enter into WE's commands. The most common use is clearing floating trees, particularly MoreTrees. Some of those trees are made up of 3 to 4 different node types. I set the WE region, target a leaf node and enter the modname:nodename to be replaced by air. Then repeat the process for the other leaf type, fruit/nut, and trunk. Inspector mods are nice but they get in the way of the work-flow for this particular use case.

Minetest is very geeky and, eventually, new comers will run into situations where they will need to know what the name of a mod is and what the name of a node is. A quick way to teach them is to tell them how to turn on the debug display, point at the ground, and report back what they see in the debug display.

**Cardinal Direction**
This was more helpful before the mini-map was added. These days it is more of a nicety for us old timers who still prefer analog clocks and mercury thermometers. Otherwise, promoting MT as educational can be demonstrated by teaching kids the cardinal directions, learning what "N", "S", "E", "W", and their diagonal variants indicate, how they relate in the min-map, and how the kid's position in the map relates to other places and landmarks.

**Coords**
For years I have recommended to players to keep pen and paper next to their computer and to jot down notes, doodle ideas, and chiefly to write down the coords to places important to them. Even with the teleporter mods we have these days, keeping note of coords still proves handy for places players have scouted but don't have a teleport too so as not to be found to easily by others or, as is more common, they mistakenly clicked on the sethome button instead of the returnhome button and need an admin or moderator to manually teleport them home.

LinuxGaming2 uses ShadowNinja's Areas protection mod. Players are able to set protection on their own stuff using chat commands. Those chat command depend on knowing the coords of two, cubically opposite corners. Another "MT is educational" example, use simple math to calculate where the second position is based on the coords of the first position.

Using coords to calculate the number of blocks required to build a wall is helpful; especially when using expensive nodes to build with.

Another common thing players have used coords for is to tell me the location of something they want me to see or check into. They pass the coords on to me and then I teleport to those coords.

When I see something suspicious in the logs, I teleport to the coords recorded in the logs to see first hand what has happened. On more than a few occasions I enter a typo, forget a number or a negative sign and wind up in the wrong place. A comparison of the coords in the debug display to the coords in the logs shows me what I entered incorrectly.

nerzhul commented:
for me the NW, NE, SE, SW should be added, with coordinate with the minimap, it make more sense to be integrated in this "Hud" instead of a hidden feature

paramat commented:
Author isn't active, probably needs adopting.

paramat commented:
Author is not active for 3 months, controversial, difficult to get a consensus, needs a rethink if splitting information is done.
